src/peft/tuners/lora/layer.py

```diff
@@ -250,6 +253,56 @@ def pissa_init(self, adapter_name, init_lora_weights):
         weight = transpose(weight.to(dtype), self.fan_in_fan_out)
         self.get_base_layer().weight.data = weight
 
+    def corda_init(self, adapter_name, init_lora_weights):
+        linear = self.get_base_layer()
+        weight = linear.weight
+        dtype = weight.dtype
+        if dtype not in [torch.float32, torch.float16, torch.bfloat16]:
+            raise TypeError(
+                "Please initialize CorDA under float32, float16, or bfloat16. "
+                "Subsequently, re-quantize the residual model to help minimize quantization errors."
+            )
+        weight = weight.to(torch.float32)
+        out_dim = weight.data.size(0)
+        in_dim = weight.data.size(1)
+
+        # Calculate WC from covariance matrix
+        if not hasattr(linear, "eigens"):
+            raise ValueError("eigens attribute not found, it's expected to be set by the pre-injection hook.")
+        eigens = linear.eigens
+        U = eigens.U_WC
+        S = eigens.S_WC
+        V = eigens.V_WC
+        r = self.r[adapter_name]
+
+        # nan or inf check
+        if torch.isnan(S).any() or torch.isinf(S).any():
+            raise ValueError("nan or inf in S")
+        if torch.isnan(U).any() or torch.isinf(U).any():
+            raise ValueError("nan or inf in U")
+        if torch.isnan(V).any() or torch.isinf(V).any():
+            raise ValueError("nan or inf in V")
+
+        # Sanity check
+        if U.size(0) != out_dim or U.size(1) != r:
+            raise ValueError(f"U size mismatch: {U.size()} vs. ({out_dim}, {r})")
+        if S.size(0) != r:
+            raise ValueError(f"S size mismatch: {S.size()} vs. ({r},)")
+        if V.size(0) != in_dim or V.size(1) != r:
+            raise ValueError(f"V size mismatch: {V.size()} vs. ({in_dim}, {r})")
+
+        # Apply alpha
+        S /= self.scaling[adapter_name]
+
+        # Init lora_A and lora_B weights
+        lora_A = V.t().mul(S.sqrt().view(-1, 1)).contiguous()
+        lora_B = U.mul(S.sqrt()).contiguous()
+        self.lora_A[adapter_name].weight.data = lora_A
+        self.lora_B[adapter_name].weight.data = lora_B
+        weight = weight.data - self.scaling[adapter_name] * lora_B @ lora_A
+        weight = weight.to(dtype)
+        self.get_base_layer().weight.data = weight
```

Comment:
Here we modify the base weights. This means that if a user uses this CorDA adapter and then loads another LoRA adapter trained on the same base model, that LoRA adapter will not work correctly. This is the same issue we have with PiSSA and OLoRA. To address this, these methods implement a way to convert the adapter into an ordinary LoRA adapter that works with the unmodified base weights, see e.g. here: https://github.com/huggingface/peft/tree/main/examples/pissa_finetuning#convert-pissa-to-lora

Do you think a similar feature could be implemented for CorDA? If this is not planned, we need to document that it's not advisable to use CorDA with other LoRA adapters, and we should probably add a check and emit a warning if we detect this.

Reply:
Yes, you are right! It also works for CorDA in the same way as PiSSA and OLoRA (double the lora rank). We are integrating this feature.

Reply:
Nice, this will be very useful to have.